fix(client-engine-runtime): cancel transactions that are still starting

aqrln opened the pull request:
> **Scope changed.** This PR originally plumbed an `AbortSignal` through `SqlDriverAdapter.startTransaction` into every adapter, to stop a timed-out start from leaking its connection. In the months since, `main` fixed that same leak by a different route — #28831 and #29727 let the start finish in the background, then send an explicit `ROLLBACK` and release the connection. Rebasing the original design would have re-opened a settled decision and risked regressing those fixes, so the PR has been reduced to the one problem `main` still has. The adapter interface is untouched.

## The bug

A transaction is only put into `transactions` once the driver hands it over. Until then it is unreachable, so `cancelAllTransactions` cannot see it:

```ts
transaction.transaction = await Promise.race([...])
this.transactions.set(transaction.id, transaction)   // only now is it visible
```

`LocalExecutor.disconnect` calls `cancelAllTransactions()` and then, in a `finally`, `driverAdapter.dispose()`. Disconnecting while a transaction is being opened therefore abandons its connection: nothing rolls it back, and the adapter is disposed out from under it.

The existing background-rollback path does not cover this, because it only runs when the `maxWait` timer fires.

## The fix

- Track starts while they are in flight, before anything is awaited, so a cancellation arriving mid-start can still find them.
- `cancelAllTransactions` aborts each one and waits for it to settle, where "settled" means the discarded transaction has been rolled back and its connection released — so callers can dispose the adapter knowing nothing is still held.
- An abort landing before the driver call is made abandons the start without opening anything, rather than opening a transaction nobody will close. That window is real: `randomUUID()` awaits a dynamic import before the driver is called.
- The background rollback moves into `#discardStartedTransaction` so it can be awaited rather than only fired and forgotten. Its behaviour, including the phantom-query handling from #29727, is unchanged.

Most of the diff is reindentation from wrapping the body in `try`/`finally`. The behavioural change is the tracking set, the guard before the driver call, and the additions to `cancelAllTransactions`.

## Tests

Two tests, both verified to fail without the fix:

- `cancelAllTransactions waits for a transaction whose start is still in flight` — asserts cancellation does **not** resolve until the rollback has happened, not merely that it happens eventually.
- `cancelAllTransactions abandons a start that has not reached the driver yet` — covers the pre-driver window.

204 tests in `@prisma/client-engine-runtime` pass; build and lint clean.

## Known wrinkle

A `startTransaction()` call cancelled by `disconnect()` rejects with `TransactionStartTimeoutError`, which is the existing error for an abandoned start but reads oddly when the cause was a disconnect. Introducing a distinct error would widen the scope again, so it is left as is.

tensordreams committed:
fix(client-engine-runtime): cancel transactions that are still starting

A transaction is only registered once the driver hands it over, so one
whose `startTransaction` call is still in flight was invisible to
`cancelAllTransactions`. `LocalExecutor.disconnect` cancels and then
disposes the driver adapter, so disconnecting while a transaction was
being opened abandoned its connection.

Track starts while they are in flight. `cancelAllTransactions` now aborts
them and waits for each to settle, so it does not return until every
discarded transaction has been rolled back and its connection released.
An abort arriving before the driver call is made is also handled: the
start is abandoned without opening anything.

The bulk of the diff is reindentation from wrapping the body in
try/finally; the behavioural change is the tracking set, the guard before
the driver call, and the additions to `cancelAllTransactions`.

tensordreams committed:
fix(client-engine-runtime): avoid double rollback, report execution time

Review follow-ups.

An abort can land after the driver has already handed the transaction
over but before the awaiting continuation runs, leaving a transaction on
the wrapper. Ownership passes to the discard, which holds its own
reference, so drop the wrapper's: otherwise the close rolls back and
releases the same transaction a second time.

`startedAt` was recorded before the id was generated and the driver was
called, so the `timeTaken` reported on an execution timeout included the
time spent opening the transaction. Record it once startup is over. The
existing test expected `START_TRANSACTION_TIME + timeout + rollbackDelay`
and now expects `timeout + rollbackDelay`.

tensordreams commented on packages/client-engine-runtime/src/transaction-manager/transaction-manager.ts:
```diff
@@ -136,83 +162,119 @@ export class TransactionManager {
-                }
-              }
-            })
-            .catch((e) => debug('error in discarded transaction:', e))
+      transaction.transaction = await Promise.race([
+        startTransactionPromise.finally(() => clearTimeout(startTimer)),
+        once(abortController.signal, 'abort').then(() => undefined),
+      ])
 
-          // Call `#closeTransaction` to update internal state. It won't actually attempt
-          // to rollback the tx itself because `transaction.transaction` is undefined.
-          await this.#closeTransaction(transaction, 'timed_out')
+      this.transactions.set(transaction.id, transaction)
 
-          throw new TransactionStartTimeoutError()
-        }
+      // Transaction status might have timed out while waiting for transaction to start. => Check for it!
+      switch (transaction.status) {
+        case 'waiting':
+          if (abortController.signal.aborted) {
+            discarding = this.#discardStartedTransaction(startTransactionPromise)
+
+            // Call `#closeTransaction` to update internal state. It won't actually attempt
+            // to rollback the tx itself because `transaction.transaction` is undefined.
+            await this.#closeTransaction(transaction, 'timed_out')
+
```
Guard added in 6e40225c — `transaction.transaction = undefined` before `#closeTransaction`, so ownership passes cleanly to `#discardStartedTransaction`, which holds its own reference.

One caveat, in the interest of not overstating it: I could not construct a deterministic reproduction. In a unit test the abort branch of the race resolves in fewer microtask hops than `startTransactionPromise.catch(...).finally(...)`, so the abort tends to win and `transaction.transaction` stays `undefined`. I wrote a test that cancels exactly as the driver hands over, and it passes with *and* without the guard — so it is a regression test for "rolled back exactly once", not evidence for the guard itself. I renamed it accordingly rather than let it imply cover it does not provide.

The reasoning still holds — nothing prevents an abort landing between the race settling and the continuation running, and the existing comment ("won't actually attempt to rollback ... because `transaction.transaction` is undefined") stopped being guaranteed the moment `cancelAllTransactions` became a second abort source. Releasing a pooled connection twice is bad enough to justify one defensive line.

tensordreams commented on packages/client-engine-runtime/src/transaction-manager/transaction-manager.ts:
```diff
@@ -136,83 +162,119 @@ export class TransactionManager {
+            // to rollback the tx itself because `transaction.transaction` is undefined.
+            await this.#closeTransaction(transaction, 'timed_out')
+
+            throw new TransactionStartTimeoutError()
+          }
+
+          transaction.status = 'running'
+          // Start timeout to wait for transaction to be finished.
+          transaction.timer = this.#startTransactionTimeout(transaction.id, options.timeout)
```
Applied in 6e40225c.

Worth recording what this changed, since an existing test had the old behaviour written into it. `commitTransaction during a rollback caused by a time out ...` expected:

```ts
timeTaken: START_TRANSACTION_TIME + timeout + rollbackDelay   // 600
```

and now expects `timeout + rollbackDelay` (400). That 200 ms was the mock driver's startup time being reported as part of the transaction's own execution — so the assertion had encoded the inaccuracy rather than caught it.

Only the diagnostic is affected: the execution timer already measured from the right point, because `#startTransactionTimeout` takes its own `Date.now()` after startup completes. `startedAt` is used solely for `timeTaken` in the timeout error.

tensordreams committed:
fix(client-engine-runtime): bound the wait for discarded transactions

Waiting for an aborted start to roll back meant waiting on the driver to
settle the `startTransaction` promise it still holds. A driver stuck on a
dead connection never does, so `cancelAllTransactions` never returned and
disconnect hung. `maxWait` does not help: it releases the caller, not the
promise the cleanup is waiting on.

Bound the wait instead. The rollback still gets a chance to finish before
the adapter is disposed, but it cannot hold shutdown open indefinitely.

tensordreams commented on packages/client-engine-runtime/src/transaction-manager/transaction-manager.ts:
```diff
@@ -136,83 +162,119 @@ export class TransactionManager {
-    ])
+      // Keep a reference to the startTransaction promise so we can clean up
+      // if the timeout fires before it completes.
+      const startTransactionPromise = this.driverAdapter
+        .startTransaction(options.isolationLevel)
+        .catch(rethrowAsUserFacing)
```
Not adopted as written, but this found a real bug in my change — thank you.

**The hazard is real.** Your stated reason ("a stuck driver start can be aborted instead of holding `cancelAllTransactions()`/shutdown open") described something this PR had just introduced. I checked rather than assumed: with a driver whose `startTransaction` never settles, `cancelAllTransactions()` never resolves, so `disconnect()` hangs indefinitely. On `main` it returns immediately.

The chain is `cancelAllTransactions` → `settled` → `#discardStartedTransaction` → `await startTransactionPromise`, with nothing bounding the last step. `maxWait` does not help: it releases the *caller's* `startTransaction()`, while the driver promise the cleanup waits on stays pending.

**Fixed in 2a0156f6, by bounding the wait rather than threading the signal.** The rollback still gets its chance before disposal, but cannot hold shutdown open. Covered by a test that hangs (5 s timeout) without the bound.

**Why not the signal.** `startTransaction` is on `SqlDriverAdapter` in the published `@prisma/driver-adapter-utils`, so adding a parameter is a breaking change for third-party adapters, and honouring it means changes across all ten shipped ones. This PR was deliberately narrowed after `main` fixed the original connection leak by a different route (#28831, #29727); reinstating that design belongs in its own PR with its own review, not as a follow-up here.

That said, the signal is the better long-term answer for precisely the reason you give: a bound is a mitigation, and only driver-level cancellation actually stops a stuck start. The grace period is documented as such in the code.

tensordreams committed:
refactor(client-engine-runtime): stop discarding rejections in settleWithin

The rejection handler was unreachable: `settled` is constructed capturing
only `resolve`, and `#discardStartedTransaction` catches internally, so
neither can reject. Its only effect was to state that failures here are
expected and thrown away, which is not true.

Race the promise against the deadline instead, so a rejection propagates
to the `Promise.allSettled` in `cancelAllTransactions` rather than being
absorbed on the way.
